## Задача 1
### 1.1 
Создать новый признак price_per_sq_lot,
который будет содержать среднюю стоимость за
один кв. метр общей площади


In [38]:
import pandas as pd
houses = pd.read_csv('kc_house_data.csv')
houses['price_per_sq_lot'] = houses['price']/houses['sqft_lot']
houses

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  sqft_above  \
0             1180      5650     1.0           0     0  ...      1180.0   
1             2570      7242     2.0           0     0  ...      2170.0   
2              770     10000     1.0           0     0  ...       770.0   
3             1960      5000     1.0           0     0  ...      1050.0   
4             1680      8080     1.0           0     0  ...      1680.0   
...            ...       ...     ...         ...   ...  ...         ...   
21608         1530      1131     3.0           0     0  ...      1530.0   
21609         2310      5813     2.0           0     0  ...      2310.0   
21610         1020      1350     2.0           0     0  ...      1020.0   
21611         1600      2388     2.0           0     0  ...      1600.0   
21612         1020      1076     2.0           0     0  ...      1020.0   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0                  0      1955             0    98178  47.5112 -122.257   
1                400      1951          1991    98125  47.7210 -122.319   
2                  0      1933             0    98028  47.7379 -122.233   
3                910      1965             0    98136  47.5208 -122.393   
4                  0      1987             0    98074  47.6168 -122.045   
...              ...       ...           ...      ...      ...      ...   
21608              0      2009             0    98103  47.6993 -122.346   
21609              0      2014             0    98146  47.5107 -122.362   
21610              0      2009             0    98144  47.5944 -122.299   
21611              0      2004             0    98027  47.5345 -122.069   
21612              0      2008             0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  price_per_sq_lot  
0               1340        5650         39.274336  
1               1690        7639         74.288870  
2               2720        8062         18.000000  
3               1360        5000        120.800000  
4               1800        7503         63.118812  
...              ...         ...               ...  
21608           1530        1509        318.302387  
21609           1830        7200         68.811285  
21610           1020        2007        297.852593  
21611           1410        1287        167.504188  
21612           1020        1357        302.044610  

[21613 rows x 22 columns]

### 1.2 
Создать новый признак delta_renovated,
который будет содержать разницу в годах между
годом реновацией дома и годом постройки дома.
Если реновации дома не было, то в новом признаке поставьте 0

In [71]:
houses['delta_renovated'] = houses['yr_renovated'] - houses['yr_built']
houses.loc[houses['delta_renovated'] < 0, 'delta_renovated'] = 0
houses.head()



id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_renovated  zipcode      lat  \
0      5650     1.0           0     0  ...             0    98178  47.5112   
1      7242     2.0           0     0  ...          1991    98125  47.7210   
2     10000     1.0           0     0  ...             0    98028  47.7379   
3      5000     1.0           0     0  ...             0    98136  47.5208   
4      8080     1.0           0     0  ...             0    98074  47.6168   

      long  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0 -122.257           1340        5650         39.274336                0   
1 -122.319           1690        7639         74.288870               40   
2 -122.233           2720        8062         18.000000                0   
3 -122.393           1360        5000        120.800000                0   
4 -122.045           1800        7503         63.118812                0   

   Год продажи  Месяц продажи  
0         2014             10  
1         2014             12  
2         2015             02  
3         2014             12  
4         2015             02  

[5 rows x 25 columns]

### 1.3 
Создайте признаки года продажи, месяца продажи


In [66]:
houses['Год продажи'] = houses['date'].apply(lambda x: x[0:4])
houses['Месяц продажи'] = houses['date'].apply(lambda x: x[4:6])                 
houses.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_renovated  zipcode      lat  \
0      5650     1.0           0     0  ...             0    98178  47.5112   
1      7242     2.0           0     0  ...          1991    98125  47.7210   
2     10000     1.0           0     0  ...             0    98028  47.7379   
3      5000     1.0           0     0  ...             0    98136  47.5208   
4      8080     1.0           0     0  ...             0    98074  47.6168   

      long  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0 -122.257           1340        5650         39.274336                0   
1 -122.319           1690        7639         74.288870               40   
2 -122.233           2720        8062         18.000000                0   
3 -122.393           1360        5000        120.800000                0   
4 -122.045           1800        7503         63.118812                0   

   Год продажи  Месяц продажи  
0         2014             10  
1         2014             12  
2         2015             02  
3         2014             12  
4         2015             02  

[5 rows x 25 columns]

### 1.4 
Удалите признаки date, zipcode, lat, long

In [48]:
houses_changed = houses.drop(columns = ['date', 'zipcode', 'lat', 'long']).copy()
houses_changed


id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
0      7129300520  221900.0         3       1.00         1180      5650   
1      6414100192  538000.0         3       2.25         2570      7242   
2      5631500400  180000.0         2       1.00          770     10000   
3      2487200875  604000.0         4       3.00         1960      5000   
4      1954400510  510000.0         3       2.00         1680      8080   
...           ...       ...       ...        ...          ...       ...   
21608   263000018  360000.0         3       2.50         1530      1131   
21609  6600060120  400000.0         4       2.50         2310      5813   
21610  1523300141  402101.0         2       0.75         1020      1350   
21611   291310100  400000.0         3       2.50         1600      2388   
21612  1523300157  325000.0         2       0.75         1020      1076   

       floors  waterfront  view  condition  ...  sqft_above  sqft_basement  \
0         1.0           0     0          3  ...      1180.0              0   
1         2.0           0     0          3  ...      2170.0            400   
2         1.0           0     0          3  ...       770.0              0   
3         1.0           0     0          5  ...      1050.0            910   
4         1.0           0     0          3  ...      1680.0              0   
...       ...         ...   ...        ...  ...         ...            ...   
21608     3.0           0     0          3  ...      1530.0              0   
21609     2.0           0     0          3  ...      2310.0              0   
21610     2.0           0     0          3  ...      1020.0              0   
21611     2.0           0     0          3  ...      1600.0              0   
21612     2.0           0     0          3  ...      1020.0              0   

       yr_built  yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  \
0          1955             0           1340        5650         39.274336   
1          1951          1991           1690        7639         74.288870   
2          1933             0           2720        8062         18.000000   
3          1965             0           1360        5000        120.800000   
4          1987             0           1800        7503         63.118812   
...         ...           ...            ...         ...               ...   
21608      2009             0           1530        1509        318.302387   
21609      2014             0           1830        7200         68.811285   
21610      2009             0           1020        2007        297.852593   
21611      2004             0           1410        1287        167.504188   
21612      2008             0           1020        1357        302.044610   

       delta_renovated  Год продажи  Месяц продажи  
0                    0            0              0  
1                   40            0              0  
2                    0            0              0  
3                    0            0              0  
4                    0            0              0  
...                ...          ...            ...  
21608                0            0              0  
21609                0            0              0  
21610                0            0              0  
21611                0            0              0  
21612                0            0              0  

[21613 rows x 21 columns]

In [105]:
houses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                21613 non-null  int64  
 1   date              21613 non-null  object 
 2   price             21613 non-null  float64
 3   bedrooms          21613 non-null  int64  
 4   bathrooms         21613 non-null  float64
 5   sqft_living       21613 non-null  int64  
 6   sqft_lot          21613 non-null  int64  
 7   floors            21613 non-null  float64
 8   waterfront        21613 non-null  int64  
 9   view              21613 non-null  int64  
 10  condition         21613 non-null  int64  
 11  grade             21613 non-null  int64  
 12  sqft_above        21611 non-null  float64
 13  sqft_basement     21613 non-null  int64  
 14  yr_built          21613 non-null  int64  
 15  yr_renovated      21613 non-null  int64  
 16  zipcode           21613 non-null  int64 

## Задача 2
Создайте датафрейм с клиентами:

In [84]:
clients = pd.DataFrame({
 'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295, 939, 3840, 280, 20, 4332, 3475, 4213, 3113,
4809, 2134, 2242, 2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
 'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060, 1189000207, 6821600300, 7137950720,
9510920050, 6131600255, 5428000070, 1788800910, 8100400160, 3123049142, 6306800010, 5083000375, 7920100025,
1951600150, 809001400, 339600110, 1622049154, 1099600250, 8563000110, 2768100205, 3995700435, 8861700030,
3303980210, 7731100066, 8146100580, 825069097, 3889100029, 9524100196]
})
clients

client_id    house_id
0        1459  8965450190
1        4684  6823100225
2        3498  5104540330
3        3942  2131701075
4        4535  1522700060
5        2308  1189000207
6        2866  6821600300
7        2765  7137950720
8        1472  9510920050
9        4236  6131600255
10       2295  5428000070
11        939  1788800910
12       3840  8100400160
13        280  3123049142
14         20  6306800010
15       4332  5083000375
16       3475  7920100025
17       4213  1951600150
18       3113   809001400
19       4809   339600110
20       2134  1622049154
21       2242  1099600250
22       2068  8563000110
23       4929  2768100205
24       1384  3995700435
25       1589  8861700030
26       3317  3303980210
27       2260  7731100066
28       1727  8146100580
29       1764   825069097
30       1611  3889100029
31       1474  9524100196

### 2.1 
Присоедините к таблице clients данные по домам через метод join

In [90]:
clients.set_index('house_id').join(houses.set_index('id'))

client_id             date      price  bedrooms  bathrooms  \
339600110        4809  20140923T000000   395000.0         3       2.50   
809001400        3113  20140922T000000   925000.0         3       1.00   
825069097        1764  20140619T000000   770000.0         3       2.50   
1099600250       2242  20141202T000000   260000.0         3       1.75   
1189000207       2308  20141021T000000   387000.0         2       2.50   
1522700060       4535  20140624T000000   518000.0         4       2.75   
1622049154       2134  20141215T000000   289900.0         3       1.75   
1788800910        939  20141020T000000   190000.0         3       1.00   
1951600150       4213  20150408T000000   180000.0         3       1.00   
2131701075       3942  20141204T000000   420000.0         3       1.75   
2768100205       4929  20140625T000000   519000.0         4       2.50   
3123049142        280  20140805T000000   452000.0         3       2.25   
3303980210       3317  20150427T000000  1115000.0         4       3.75   
3889100029       1611  20140617T000000   810000.0         3       2.50   
3995700435       1384  20140702T000000   265000.0         4       3.00   
5083000375       4332  20141027T000000   170000.0         3       1.00   
5083000375       4332  20150319T000000   235000.0         3       1.00   
5104540330       3498  20150508T000000   679000.0         4       2.50   
5428000070       2295  20150511T000000   770000.0         5       3.50   
6131600255       4236  20141222T000000   202500.0         3       2.00   
6306800010         20  20140811T000000   436472.0         4       2.50   
6821600300       2866  20150318T000000   886000.0         3       2.25   
6823100225       4684  20150414T000000   700000.0         4       1.75   
7137950720       2765  20150304T000000   339100.0         4       2.50   
7731100066       2260  20140612T000000   545000.0         3       1.00   
7920100025       3475  20150427T000000   450000.0         2       1.00   
8100400160       3840  20150413T000000   700000.0         3       2.25   
8146100580       1727  20141020T000000   765000.0         3       1.00   
8563000110       2068  20150424T000000   427000.0         4       1.75   
8861700030       1589  20141212T000000   510000.0         3       1.50   
8965450190       1459  20150218T000000   295000.0         3       2.50   
9510920050       1472  20140902T000000   725000.0         3       2.50   
9524100196       1474  20141117T000000   239000.0         2       1.50   

            sqft_living  sqft_lot  floors  waterfront  view  ...  \
339600110          1610      3755     2.0           0     0  ...   
809001400          1630      3200     1.5           0     0  ...   
825069097          2650     40705     2.0           0     0  ...   
1099600250         1710      6400     1.0           0     0  ...   
1189000207         1170      1394     2.0           0     0  ...   
1522700060         2520     14021     2.0           0     0  ...   
1622049154         1899     11325     2.0           0     0  ...   
1788800910         1200     10458     1.0           0     0  ...   
1951600150         1610      8500     1.5           0     0  ...   
2131701075         1720      5000     1.5           0     0  ...   
2768100205         1950      2617     1.5           0     0  ...   
3123049142         2600     14810     1.0           0     2  ...   
3303980210         4040     14212     2.0           0     0  ...   
3889100029         2670     10481     2.0           0     0  ...   
3995700435         1940      8170     1.0           0     0  ...   
5083000375         1310      9529     1.0           0     0  ...   
5083000375         1310      9529     1.0           0     0  ...   
5104540330         3680      7236     2.0           0     0  ...   
5428000070         4750      8234     2.0           0     2  ...   
6131600255         1540      8316     1.0           0     0  ...   
6306800010         2692      8392     2.0           0     0  ...   

### 2.2 
Присоедините к таблице clients данные по домам через метод merge

In [92]:
clients.merge(houses, left_on = 'house_id', right_on='id').head()

client_id    house_id          id             date     price  bedrooms  \
0       1459  8965450190  8965450190  20150218T000000  295000.0         3   
1       4684  6823100225  6823100225  20150414T000000  700000.0         4   
2       3498  5104540330  5104540330  20150508T000000  679000.0         4   
3       3942  2131701075  2131701075  20141204T000000  420000.0         3   
4       4535  1522700060  1522700060  20140624T000000  518000.0         4   

   bathrooms  sqft_living  sqft_lot  floors  ...  yr_renovated  zipcode  \
0       2.50         1500      3060     2.0  ...             0    98006   
1       1.75         1870      6000     1.0  ...             0    98199   
2       2.50         3680      7236     2.0  ...             0    98038   
3       1.75         1720      5000     1.5  ...          2009    98019   
4       2.75         2520     14021     2.0  ...             0    98019   

       lat     long  sqft_living15  sqft_lot15  price_per_sq_lot  \
0  47.5605 -122.117           2700        7734         96.405229   
1  47.6435 -122.399           1710        6000        116.666667   
2  47.3543 -122.005           3310        7180         93.836374   
3  47.7380 -121.983           1410        8300         84.000000   
4  47.7344 -121.957           2330       14007         36.944583   

   delta_renovated  Год продажи  Месяц продажи  
0                0         2015             02  
1                0         2015             04  
2                0         2015             05  
3               77         2014             12  
4                0         2014             06  

[5 rows x 27 columns]

## Задача 3
### 3.1 
Найдите среднюю стоимость домов в
зависимости от количества спален
Отсортируйте от меньшей стоимости к большей


In [115]:
cost_by_bedrooms = houses.groupby('bedrooms').agg({'price': 'mean'})
cost_by_bedrooms['price'] = cost_by_bedrooms['price'].astype(int)
cost_by_bedrooms.sort_values('price')

price
bedrooms         
1          317642
2          401372
0          409503
3          466232
11         520000
4          635419
33         640000
5          786599
10         819333
6          825520
9          893999
7          951184
8         1105076

### 3.2 
Найдите минимальную, среднюю и
максимальную стоимости домов в зависимости от
состояния дома

In [116]:
cost_by_condition = houses.groupby('condition').agg({'price': ['min', 'max','mean']})
cost_by_condition

price                          
                min        max           mean
condition                                    
1           78000.0  1500000.0  334431.666667
2           80000.0  2555000.0  327287.145349
3           75000.0  7062500.0  542012.578148
4           89000.0  7700000.0  521200.390033
5          110000.0  3650000.0  612418.089359

### 3.3 
Постройте таблицу с подсчетом количества
домов в данных в зависимости от вида на
набережную и оценкой вида

In [118]:
pd.crosstab(index=houses['waterfront'], columns=houses['view'])

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

### 3.4 
Каких домов в зависимости от этажности и
количества спален больше?


In [164]:
pd.crosstab(index=houses['floors'], columns=houses['bedrooms'])

bedrooms  0    1     2     3     4    5    6   7   8   9   10  11  33
floors                                                               
1.0        4  162  1951  5455  2383  605  104   9   5   0   1   0   1
1.5        0   21   182   786   698  185   30   7   1   0   0   0   0
2.0        6   12   497  3118  3682  775  119  19   6   4   2   1   0
2.5        0    1     5    56    58   23   14   2   0   2   0   0   0
3.0        2    3   123   405    61   13    5   1   0   0   0   0   0
3.5        1    0     2     4     0    0    0   0   1   0   0   0   0

Больше всего домов с одним этажом и тремя спальнями (5455 шт)

### 3.5 
Постройте таблицу с подсчетом медианной
стоимости домов в данных в зависимости от
состояния дома и оценки дома

In [169]:
pd.crosstab(index=houses['condition'], columns=houses['grade'], values=houses['price'], aggfunc='mean').fillna('')

grade            1         3         4              5              6   \
condition                                                               
1          142000.0            150000.0  280833.333333  303813.636364   
2                    280000.0  179600.0  206166.666667  269483.898305   
3                     75000.0  210000.0  246993.460000  290651.847343   
4                              243905.0  245299.011905  318731.570803   
5                    262000.0            271127.588235  310140.858871   

grade                 7              8               9               10  \
condition                                                                 
1          334416.666667  932500.000000                                   
2          342698.520000  421996.153846        715000.0       1752500.0   
3          385666.785823  510102.771609   726587.798138  1012934.128122   
4          412923.073420  589527.303443   889664.524664  1176284.589744   
5          479667.746699  736538.048718  1123419.896825  1735830.363636   

grade                  11              12              13  
condition                                                  
1                                                          
2                                                          
3          1414401.442771  2138601.081081  3338636.363636  
4          1839916.428571  2467961.538462       5750000.0  
5          2238477.272727       2290000.0